### ISSUE TITLE
If unforgeable, might help solve Jessie this-capture issue.

### ISSUE THREAD

**Issue by erights** [hax/proposal-function-this#1]
See the section "Avoid this-capture Hazards" at https://github.com/Agoric/Jessie#avoid-this-capture-hazards

If we had an unforgeable test that a function cannot access its `this` argument, then we could consider other approaches to this this-capture hazard. For example, if we had a membrane between SES and Jessie, it could filter out this-sensitive SES functions from reaching Jessie code.

Attn @michaelfig

**Comment by ljharb** [comment]
Unforgeable, however, means it really can’t be a data property, it’d have to be an internal slot or similar, created/installed at the same time as the function was created, with a static or syntactic method of testing any function.

**Comment by hax** [comment]
I am not sure whether I understand this issue. If SES make `thisArgumentExpected` unconfigurable, would it be "unforgeable"?

**Comment by ljharb** [comment]
I'm not sure how it would be practical for SES to make every function have a nonconfigurable own data property; and either way, it would have to have that property in the absence of SES.

Additionally, even with SES and a nonconfigurable data property, a Proxy around a function could still forge any value it liked for that property.

**Comment by erights** [comment]
I agree that it should not be a data property. Whatever the test is, it must not be fooled by proxies, or it would not provide any safety against this-capture. Simply punching through to the target wouldn't be adequate because of the `apply` trap. Not clear what to do here.

**Comment by hax** [comment]
This proposal intentionally use configurable data property to satisfy the edge cases described in the README. But when I design the feature, I also considered some function reflection APIs (something like `Function.reflect(func)`) and could be a separate proposal. Would such API help for SES case?

**Comment by erights** [comment]
>  I also considered some function reflection APIs (something like `Function.reflect(func))` and could be a separate proposal. Would such API help for SES case?

Yes. Say it is something like `Function.isThisInsensitive(func)`. In SES the `Function` constructor is frozen and so would be reliable. We'd still need to figure out what `isThisInsensitive` does with a function proxy. I don't have a good answer for that one. Security demands that the `isThisInsensitive` never claim a proxy is insensitive to `this` if the proxy can be sensitive to `this`. Transparency demands that a membrane proxy around a function (a wet proxy for a dry function) act as much like the function as possible. These two conflict in ways I don't know how to resolve.

**Comment by ljharb** [comment]
Presumably the this-sensitivity of the `apply` handler would dictate the value for the proxy.

**Comment by erights** [comment]
If the proxy read the handlers on proxy initialization, remembered them, and reused them with no substitution allowed. It was a mistake that proxies were not designed to do that. (Documented in an issue somewhere.)

However, the `apply` handler is gotten each time from the handler object with a [[Get]]. If the `apply` property of the handler were a non-writable, non-configurable data property, then it would be guaranteed to have a stable value. But it seems bizarre, and an encapsulation break, for an `isThisInsensitive(aProxy)` test to punch through the proxy to test a property of the proxy's handler.

Having an `isThisInsensitive` test would be very useful. But this way of achieving it seems too ugly. Can some other way be found?

**Comment by erights** [comment]
> It was a mistake that proxies were not designed to do that.

Sorry for the passive voice when I am partly to blame.

**Comment by ljharb** [comment]
I think the problem with such a test is that it can’t easily be infallible. What’s the this-sensitivity of the following?:
```js
function f() { return this; } // presumably, true
f.bind() // ?
f.bind(null) // ?
Function.call.bind(f) // ?
Function.call.bind(f, null) // ?
function () { return f.call(this); }.bind() // ?
```
Now repeat the exercise for `f` but using `Function` or `eval` in creative ways to hide its sensitivity, or to forge it, or indeed with a Proxy around f, or with a call or bind from different realms, etc.

tbh I’m not sure how we’d spec those answers, or reason about them reliably in a general way.

**Comment by erights** [comment]
```js
function f() { return this; } // sensitive
f.bind() // insensitive
f.bind(null) // insensitive
Function.call.bind(f) // insensitive
Function.call.bind(f, null) // insensitive
function () { return f.call(this); }.bind() // insensitive
```

I didn't test, but my understanding is that the result of `bind`, a bound function, is never this-sensitive. Do you have a counter-example?

**Comment by erights** [comment]
> Now repeat the exercise for f but using Function or eval in creative ways to hide its sensitivity, or to forge it, or indeed with a Proxy around f, or with a call or bind from different realms, etc.

I don't expect any of these to be problematic except for proxy. Can you give an example of a non-proxy problematic case?

**Comment by ljharb** [comment]
```js
new (f.bind()) // sensitive, because it has a `this`
```

**Comment by erights** [comment]
> ```js
> new (f.bind()) // sensitive, because it has a `this`
> ```

Ah! I should have chosen clearer words. It is not sensitive to the `this`-argument passed in. Thus, it does not create the dangers explained at https://github.com/Agoric/Jessie#avoid-this-capture-hazards
I don't care if it creates a new object that, within itself, it calls `this`. It ignores any `this`-argument coming from the caller.
